include/range/v3/detail/optional.hpp

```diff
@@ -86,6 +86,49 @@ namespace ranges
             }
         };
 
+        template<typename T>
+        struct optional<T &>
+        {
+        private:
+            variant<meta::nil_, T &> data_;
```

Comment:
After which I *think* you get no benefit from specializing `semiregular` for references over simply using the base template.